Create NOTICE.txt

jasonblais opened the pull request:
Before submitting, please confirm you've
 - [X] read and understood our [Contributing Guidelines](https://github.com/mattermost/desktop/blob/master/CONTRIBUTING.md)
 - [X] completed [Mattermost Contributor Agreement](http://www.mattermost.org/mattermost-contributor-agreement/)
 - [x] executed `npm run lint:js` for proper code formatting

Please provide the following information:

**Summary**
Create NOTICE.txt

**Issue link**
Pre-requisite for inline spellchecker: https://github.com/mattermost/desktop/issues/225

**Test Cases**
N/A

**Additional Notes**
This is still a work-in-progress with a few additional steps
 - [x] @yuya-oc verify the source (github link) and description included for each dependency is correct
   - [x] @jasonblais update based on feedback
 - [x] @yuya-oc list any libraries that were modified. I wrote this document assuming none of them were
   - [x] @jasonblais properly attribute dependencies of `src/package.json` that were combined into a single JavaScript file by using webpack and hence modified (although the original dependencies weren't)
 - [x] @yuya-oc remove/replace the `chai-as-promised` dependency, since its license isn't acceptable: https://github.com/mattermost/desktop/pull/502
 - [x] @jasonblais confirm proper attribution of react libraries
 - [x] @jasonblais confirm proper attribution of babel libraries
 - [x] @jasonblais confirm proper attribution of electron-builder libraries
 - [x] @jasonblais add [`url-loaded` package](https://www.npmjs.com/package/url-loader)
 - [ ] @jasonblais ask for final review from it33

yuya-oc commented:
Thanks, I will check.

Not urgent, but we should automate this process to maintain `NOTICE.txt` in future. Would you explain how to write the file?

> list any libraries that were modified. I wrote this document assuming none of them were

Now all dependencies are used as is. But dependencies of `src/package.json` are combined into a single JavaScript file by using webpack. This means original files are modified though, what about this?

jasonblais commented:
> Not urgent, but we should automate this process to maintain NOTICE.txt in future. Would you explain how to write the file?

I just wrote it manually --

I don't think we need to automate it as we only need to update NOTICE.txt when we add/remove a dependency.

> Now all dependencies are used as is. But dependencies of src/package.json are combined into a single JavaScript file by using webpack. This means original files are modified though, what about this?

Awesome, that's useful to know. I added a follow-up task to make the necessary modifications.

yuya-oc commented on NOTICE.txt:
```diff
@@ -0,0 +1,1052 @@
+
+## eslint-plugin-react
+
+This product contains 'eslint-plugin-react', a tool for React specific linting rules for ESLint, by Yannick Croissant.
```
~a tool for~~ React specific linting rules for ESLint

yuya-oc commented on NOTICE.txt:
```diff
@@ -0,0 +1,1052 @@
+
+## webpack-dev-server
+
+This product contains 'webpack-dev-server', which provides fast in-memory access to the webpack asset, by Webpack.
```
which serves a webpack app

yuya-oc commented:
> I don't think we need to automate it as we only need to update NOTICE.txt when we add/remove a dependency, which happens quite rarely.

I understood by reading NOTICE.txt. It seems hard to automate.

jasonblais commented:
Thanks! Updated with the following two commits:
 - https://github.com/mattermost/desktop/pull/488/commits/acedeb4f94de8c51d785c053180efc72fa93fe76
 - https://github.com/mattermost/desktop/pull/488/commits/7baf5c9542ce5874e5b18a6a120cce3f1570d012

jasonblais commented:
In your comment here: https://github.com/mattermost/desktop/pull/488#issuecomment-289212099

> Now all dependencies are used as is. But dependencies of src/package.json are combined into a single JavaScript file by using webpack. This means original files are modified though, what about this?

Was the only modification that they were combined into a single JavaScript file, but otherwise no change?

yuya-oc commented on NOTICE.txt:
```diff
@@ -1050,3 +1050,120 @@ NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
+## babel
+
+This product contains the following packages of 'babel', a compiler for writing next generation JavaScript: 'babel-core', a Babel compiler core, 'babel-plugin-transform-object-rest-spread', a plugin that allows Babel to transform rest properties for object destructuring assignment and spread properties for object literals, and 'babel-preset-react', a Babel preset for all React plugins.
+'react-addons-css-transition-group', by Babel.
```
`react-addons-css-transition-group` is an addon of React. It's already written at React section.

jasonblais replied:
sorry, copy-paste error :/

yuya-oc commented on NOTICE.txt:
```diff
@@ -1167,3 +1167,35 @@ NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
+
+## url-loader
+
+This product contains the following packages of 'url-loader', a url loader module for webpack, by Webpack.
```
by Webpack community. (see https://github.com/webpack-contrib)

yuya-oc commented on NOTICE.txt:
```diff
@@ -0,0 +1,1201 @@
+## underscore
+
+This product contains 'underscore', a utility-belt library for JavaScript that provides support for usual functional suspects
+without extending any core JavaScript objects, by Jeremy Ashkenas, DocumentCloud and Investigative.
```
Probably "Jeremy Ashkenas, DocumentCloud and Investigative Reporters & Editors".

yuya-oc commented on NOTICE.txt:
```diff
@@ -0,0 +1,1201 @@
+
+## url-loader
+
+This product contains the following packages of 'url-loader', a url loader module for webpack, by Webpack.
```
This product contains 'url-loader', ...

yuya-oc commented on NOTICE.txt:
```diff
@@ -0,0 +1,1201 @@
+
+## babel-loader
+
+This product contains the following packages of 'babel-loader', a webpack plugin for Babel, by Babel.
```
This product contains 'babel-loader', ...

yuya-oc commented on NOTICE.txt:
```diff
@@ -0,0 +1,1201 @@
+
+## babel-eslint
+
+This product contains the following packages of 'babel-eslint', an ESLint using Babel as the parser, by Babel.
```
This product contains 'babel-eslint', ...

jasonblais replied:
and here too.

yuya-oc commented:
> Was the only modification that they were combined into a single JavaScript file, but otherwise no change?

I believe yes.

Just for note. If #387 was applied, combined files are definitely modified. On the other hand, `simple-spellchecker` is never modified because it's not combined into a single JS. It's not adapted to webpack (see [#277](https://github.com/mattermost/desktop/pull/277#issue-176242982)).

jasonblais commented:
Fixed the typos and added a `contains a modified portion of...` for libraries in `src/package.json`. 

I have a task for myself to create a separate attribution list on how they were modified (were combined into a single JavaScript file by using webpack and hence modified. Plus perhaps further changes later with #387), but this will be sufficient for now and is how it's handled in other mattermost repos.

jasonblais commented:
Let me know if this looks okay @yuya-oc and I can share with Ian for final review.

jasonblais commented:
Awesome! I've queued it for my next meeting with Ian on Thursday

jasonblais commented:
Approved by Ian
